tensorflow_datasets/text/pg19.py

```diff
@@ -0,0 +1,124 @@
+
+_CITATION = """
+@article{raecompressive2019,
+author = {Rae, Jack W and Potapenko, Anna and Jayakumar, Siddhant M and
+          Hillier, Chloe and Lillicrap, Timothy P},
+title = {Compressive Transformers for Long-Range Sequence Modelling},
+journal = {arXiv preprint},
+url = {https://arxiv.org/abs/1911.05507},
+year = {2019},
+}
+"""
+
+_DESCRIPTION = """
+This dataset contains the PG-19 language modeling benchmark. It includes a set
+of books extracted from the Project Gutenberg books project
+(https://www.gutenberg.org), that were published before 1919. It also contains
+metadata of book titles and publication dates.
+PG-19 is over double the size of the Billion Word benchmark and contains
+documents that are 20X longer, on average, than the WikiText long-range
+language modelling benchmark.
+
+Books are partitioned into a train, validation, and test set. Books metadata is
+stored in metadata.csv which contains
+(book_id, short_book_title, publication_date, book_link).
+"""
+
+_DATA_DIR = 'gs://deepmind-gutenberg'
+
+class Pg19(tfds.core.GeneratorBasedBuilder):
+  """This dataset contains the PG-19 language modeling benchmark"""
+
+  VERSION = tfds.core.Version('0.1.0')
+
+  def _info(self):
+
+    return tfds.core.DatasetInfo(
+        builder=self,
+        description=_DESCRIPTION,
+        features=tfds.features.FeaturesDict({
+            'book_text': tfds.features.Text(),
+            'book_id': tf.int32,
+            'book_title': tf.string,
+            'publication_date': tf.string,
+            'book_link': tf.string
+        }),
+        supervised_keys=None,
+        homepage='https://github.com/deepmind/pg19',
+        citation=_CITATION,
+    )
+
+  def _split_generators(self, dl_manager):
+    """Returns SplitGenerators."""
+    del dl_manager
+
+    metadata_dict = dict()
+    metadata_path = os.path.join(_DATA_DIR, "metadata.csv")
+    metadata = tf.io.gfile.GFile(metadata_path).read().splitlines()
+
+    for row in metadata:
+      row_split = row.split(',')
+      # book_id: [book_title, publication_date, book_link]
+      metadata_dict[int(row_split[0])] = row_split[1:]
+
+    return [
+        tfds.core.SplitGenerator(
+            name=tfds.Split.TRAIN,
+            gen_kwargs={
+                "metadata": metadata_dict,
+                "filepath": os.path.join(_DATA_DIR, "train")},
+        ),
+        tfds.core.SplitGenerator(
+            name=tfds.Split.VALIDATION,
+            gen_kwargs={
+                "metadata": metadata_dict,
+                "filepath": os.path.join(_DATA_DIR, "validation")},
+        ),
+        tfds.core.SplitGenerator(
+            name=tfds.Split.TEST,
+            gen_kwargs={
+                "metadata": metadata_dict,
+                "filepath": os.path.join(_DATA_DIR, "test")},
+        ),
+    ]
+
+  def _generate_examples(self, filepath, metadata):
+    """Yields examples."""
+    for index, file in enumerate(tf.io.gfile.listdir(filepath)):
```

Comment:
`listdir` is not deterministic, so generating the dataset twice on different system may lead different keys for the same examples. Instead, you could use `book_id` or similar

Reply:
Or just enumerate(sorted(...)) ?

Reply:
Why do you need it in sorted order?
we can do something like  this 
` enumerate(sorted(tf.io.gfile.listdir(path), key=lambda l: int(l.rstrip('.txt'))))`
any thoughts?